lib/batcher.mli

```diff
@@ -1,34 +1,93 @@
-module type BatchedDS =
-sig
+module type S = sig
   type t
-  (** Type of the Explicitly Batched Data Structure *)
+  (** [t] represents a vanilla data structure. *)
 
-  type 'a batch_op 
-  (** GADT specifiying the metadata of the operation to be performed including the parameters of the operation exepected *)
+  type 'a op 
+  (** ['a op] represents a single operation on [t] with the return type ['a]. *)
 
-  (* Change constructor to Wrapped *)
-  type wrapped_batch_op =
-      Batched_op : 'a batch_op * ('a -> unit) -> wrapped_batch_op
-    (** Wrapper to make 'a batch_op an existential type so that we can batch different operations together and pass them around easily *)
+  type wrapped_op = Mk : 'a op * ('a -> unit) -> wrapped_op
+  (** [wrapped_op] represents an operation on the datastructure and
+      the continuation to run after its completion.  *)
 
-  val create : unit -> t
-  (** [create pool] returns a new instance of the batched data structure. holding a reference to the pool to provide dynamic multi-threading functionality *)
+  val init : unit -> t
+  (** [init ()] returns a new instance of the data structure. *)
 
-  val bop : t -> Task.pool -> wrapped_batch_op array -> int -> unit
-  (** [bop t pool op_array num] accepts an `op_array` containing `num` arbritrary wrapped_batch_op and performs batch-parallel operations to completion *)
+  val run : t -> Task.pool -> wrapped_op array -> unit
+  (** [run t pool ops num] when called with a data structure [t], and
+      a thread pool [pool], executes all the operations in [ops],
+      possibly using parallelism to improve the speed of the
+      operation. *)
 
 end
-module Make :
-  functor (DS : BatchedDS) ->
-  sig
-    type t
-    type 'a batch_op = 'a DS.batch_op
-    val create : Task.pool -> t
-    (* Change to name to apply *)
-    val batchify : t -> 'a batch_op -> 'a
-
-    val unload : t -> DS.t
-    [@@@alert unsafe "For developer use"]
-    val load : t -> DS.t -> unit
-    [@@@alert unsafe "For developer use"]
-  end
\ No newline at end of file
+
+module type S1 = sig
+  type 'a t
+  (** ['a t] represents a data structure parameterised over ['a]. *)
+
+  type ('a, 'b) op 
+  (** [('a, 'b) op] represents a single operation on ['a t] with the return type ['b]. *)
+
+  type 'a wrapped_op = Mk : ('a, 'b) op * ('b -> unit) -> 'a wrapped_op
+  (** ['a wrapped_op] represents an operation on the datastructure ['a
+      t] and the continuation to run after its completion.  *)
+
+  val init : unit -> 'a t
+  (** [init ()] returns a new instance of the data structure. *)
+
+  val run : 'a t -> Task.pool -> 'a wrapped_op array -> unit
+  (** [run t pool ops num] when called with a data structure ['a t], and
+      a thread pool [pool], executes all the operations in [ops],
+      possibly using parallelism to improve the speed of the
+      operation. *)
+
+end
```

Comment:
Why did you make the type of the BatchedDS polymorphic? I thought we agreed to keep them abstract and provide the DS's as functors

Reply:
After starting to refactor your experiments code, I realised that you were right, and my  initial guess was wrong - it really does make sense to have a separate polymorphic interface -- hence the S1 and Make1 functors.